circularqueue_test.go

```diff
@@ -0,0 +1,180 @@
+			expectedPeek: [][]circularInt{{6, 7, 8}, {7, 8, 9, 10, 11}},
+			add:          [][]circularInt{{8}, {9, 10, 11}},
+			rem:          []int{5, 1},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove then add 3",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12},
+			expectedPeek: [][]circularInt{{6, 7, 8}, {}, {}},
+			add:          [][]circularInt{{8}, {9, 10, 11}, {12}},
+			rem:          []int{5, 6, 1},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove too many too early then add one in",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11},
+			expectedPeek: [][]circularInt{{6, 7, 8}, {}, {12}},
+			add:          [][]circularInt{{8}, {9, 10, 11}, {12}},
+			rem:          []int{5, 33, 0},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove everyone",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			expectedPeek: [][]circularInt{{6, 7}, {}},
+			add:          [][]circularInt{{}, {}},
+			rem:          []int{5, 2},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove everything but one then fill it up again",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			expectedPeek: [][]circularInt{{6, 7}, {}, {8, 9, 10, 11}},
+			add:          [][]circularInt{{}, {}, {8, 9, 10, 11}},
+			rem:          []int{5, 2, 0},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			q := newCircularIntCircularQueue(c.starting...)
+			//fmt.Println(q.data)
+			res := []circularInt{}
+			for j := 0; j < len(c.rem); j++ {
+				q.Enqueue(c.add[j]...)
+
+				for i := 0; i < c.rem[j] && q.Next(); i++ {
+					res = append(res, q.Val())
+				}
+				peekRes := []circularInt{}
+				p, ok := q.Peek(0)
+				for i := 1; ok; i++ {
+					peekRes = append(peekRes, p)
+					p, ok = q.Peek(i)
+				}
+				if !reflect.DeepEqual(peekRes, c.expectedPeek[j]) {
+					t.Errorf("expected peeking, on step %d we would see %v, got %v", j, c.expectedPeek[j], peekRes)
+
+				}
+			}
+			if !reflect.DeepEqual(res, c.expected) {
+				t.Errorf("expected %v, got %v", c.expected, res)
+			}
+		})
+
+	}
+}
+
+func Test_leakCircularBuf(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Testing for leaks can be slow, because of the way finalizers work")
+	}
+	finalizedItems := []int{} // this can't be pointers because we need the objects to leave memory
+	finalizedLock := &sync.Mutex{}
+	expectedFinalizedItems := []int{}
+	// fill the expectedFinalizedItems
+	for i := 0; i < 20; i++ {
+		expectedFinalizedItems = append(expectedFinalizedItems, i)
+	}
+
+	q := newCircularIntPtrCircularQueue(nil)
+	for i := 0; i < len(expectedFinalizedItems); i++ {
+		i := i // make i a local object
+		item := circularIntPtr(&i)
+		runtime.SetFinalizer(item, func(q circularIntPtr) {
+			// if the finalizer is called, that means that the GC believes the objects should be freed
```

Comment:
We use `runtime.SetFinalizer` to test if the GC is seeing the objects for deletion.